mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -622,29 +678,108 @@ def _compare_metrics(metric_a, metric_b):
+            old_metric.step = new_metric.step
+            old_metric.timestamp = new_metric.timestamp
+            old_metric.is_nan = new_metric.is_nan
+            return old_metric
+
+        if not logged_metrics:
+            return
+
+        # Fetch the latest metric value corresponding to the specified run_id and metric keys and
+        # lock their associated rows for the remainder of the transaction in order to ensure
         # isolation
-        latest_metric = (
-            session.query(SqlLatestMetric)
-            .filter(
-                SqlLatestMetric.run_uuid == logged_metric.run_uuid,
-                SqlLatestMetric.key == logged_metric.key,
+        latest_metrics = {}
+        metric_keys = [m.key for m in logged_metrics]
+        # Divide metric keys into batches of 500 to avoid binding too many parameters to the SQL
+        # query, which may produce limit exceeded errors or poor performance on certain database
+        # platforms
+        metric_key_batches = [metric_keys[i : i + 500] for i in range(0, len(metric_keys), 500)]
+        for metric_key_batch in metric_key_batches:
+            # First, determine which metric keys are present in the database
+            latest_metrics_key_records_from_db = (
+                session.query(SqlLatestMetric.key)
+                .filter(
+                    SqlLatestMetric.run_uuid == logged_metrics[0].run_uuid,
+                    SqlLatestMetric.key.in_(metric_key_batch),
+                )
+                .all()
             )
-            .with_for_update()
-            .one_or_none()
-        )
-        if latest_metric is None or _compare_metrics(logged_metric, latest_metric):
-            session.merge(
-                SqlLatestMetric(
+            # Then, take a write lock on the rows corresponding to metric keys that are present,
+            # ensuring that they aren't modified by another transaction until they can be
+            # compared to the metric values logged by this transaction while avoiding gap locking
+            # and next-key locking which may otherwise occur when issuing a `SELECT FOR UPDATE`
+            # against nonexistent rows
+            if len(latest_metrics_key_records_from_db) > 0:
+                latest_metric_keys_from_db = [
+                    record[0] for record in latest_metrics_key_records_from_db
+                ]
+                latest_metrics_batch = (
+                    session.query(SqlLatestMetric)
+                    .filter(
+                        SqlLatestMetric.run_uuid == logged_metrics[0].run_uuid,
+                        SqlLatestMetric.key.in_(latest_metric_keys_from_db),
+                    )
+                    # Order by the metric run ID and key to ensure a consistent locking order
+                    # across transactions, reducing deadlock likelihood
+                    .order_by(SqlLatestMetric.run_uuid, SqlLatestMetric.key)
+                    .with_for_update()
```

Comment:
@erensahin After adding a concurrency test, I found that using a serializable transaction leads to deadlock due to range locking. I've instead reverted back to the previous `with_for_update()` row locking procedure and derisked it by ensuring that we only lock preexisting rows in the database, resolving the issue described in https://github.com/mlflow/mlflow/issues/4353#issue-890814942. While it's possible that two separate transactions may simultaneously insert a brand new metric, this case is highly unlikely to occur in practice, and it wasn't handled before. The slightly more common case where two threads are logging values for a preexisting metric simultaneously is safe from a concurrency perspective with row locking.